File: XIRHXQ_Line.m
classdef XIRHXQ_Line < handle
    properties
        s(1, 1) XIRHXQ_Point
        e(1, 1) XIRHXQ_Point
    end
%{
    methods of XIRHXQ_Line:
    
    - XIRHXQ_Line(start_point, end_point)
        !!! support XIRHXQ_Point type and vector type
        a line crossing start_point and end_point
    
    - XIRHXQ_Line(start_point, angle)
        a line crossing start_point with angle
    
    - XIRHXQ_Line(a, b, c)
        a line with equation ax + by + c = 0
    
    - adjust(obj)
        sort s and e such that s < e
    
    - length(obj)
        distance between start point and end point
    
    - angle(obj)
        angle of the line
    
    - relation(obj, p)
        the relation of p and the line
        -1: p is on the left/down
        1: p is on the right/up
        0: p is on the line
        
    - relation_to_seg(obj, p)
        the relation of p and the segment
        -2: p is not on the line
        -1: p is on the left/down
        1: p is on the right/up
        0: p is on the segment
    
    - is_on_line(obj, p)
        whether the point p is on the line
    
    - is_on_seg(obj, p)
        whether the point p is on the segment
    
    - parallel(obj1, obj2)
        whether obj1 is parallel to or same as obj2
    
    - is_the_same_line(obj1, obj2)
        whether obj1 is the same line as obj2
    
    - line_cross_line(obj1, obj2)
        whether line obj1 and line obj2 intersect
    
    - cross_point(obj1, obj2)
        !!! obj1 and obj2 must cross
        returns the cross_point
    
    - line_cross_seg(obj1, obj2)
        whether line obj1 and segment obj2 intersect
    
    - seg_cross_seg(obj1, obj2)
        whether segment obj1 and segment obj2 intersect
    
    - dis_point_to_line(obj, p)
        returns the distance from p to line obj
    
    - dis_point_to_seg(obj, p)
        returns the distance from p to segment obj
    
    - dis_seg_to_seg(obj, l)
        returns the distance from segment obj to segment l
    
    - project_point(obj, p)
        returns the project point of p on line obj
    
    - symmetry_point(obj, p)
        returns the symmetric point of p with respect to line obj
        
%}
    methods
        function obj = XIRHXQ_Line(varargin)
            if nargin == 2
                obj.s = XIRHXQ_Point(varargin{1});
                if isnumeric(varargin{2}) && isscalar(varargin{2})
                    angle = varargin{2};
                    obj.e = obj.s + XIRHXQ_Point(cos(angle), sin(angle));
                else
                    obj.e = XIRHXQ_Point(varargin{2});
                end
            elseif nargin == 3
                a = varargin{1};
                b = varargin{2};
                c = varargin{3};
                mustBeNumeric(a);mustBeNumeric(b);mustBeNumeric(c);
                if a == 0
                    obj.s = XIRHXQ_Point(0, -c/b);
                    obj.e = XIRHXQ_Point(1, -c/b);
                elseif b == 0
                    obj.s = XIRHXQ_Point(-c/a, 0);
                    obj.e = XIRHXQ_Point(-c/a, 1);
                else
                    obj.s = XIRHXQ_Point(0, -c/b);
                    obj.e = XIRHXQ_Point(1, -(c+a)/b);
                end
            end
            if obj.s > obj.e
                tmp = obj.s;
                obj.s = obj.e;
                obj.e = tmp;
            end
        end
        function obj = adjust(obj)
            if obj.s > obj.e
                tmp = obj.s;
                obj.s = obj.e;
                obj.e = tmp;
            end
        end
        function res = length(obj)
            res = distance(obj.s, obj.e);
        end
        function res = angle(obj)
            res = obj.s.angle(obj.e);
        end
        function res = relation(obj, p)
            p = XIRHXQ_Point(p);
            c = (p - obj.s) ^ (p - obj.e);
            if c > 0
                res = 1;
            elseif c < 0
                res = -1;
            else
                res = 0;
            end
        end
        function res = relation_to_seg(obj, p)
            p = XIRHXQ_Point(p);
            if obj.relation(p) ~= 0
                res = -2;
            elseif p > obj.e
                res = 1;
            elseif p < obj.s
                res = -1;
            else
                res = 0;
            end
         end
        function res = is_on_line(obj, p)
            p = XIRHXQ_Point(p);
            res = (obj.relation(p) == 0);
        end
        function res = is_on_seg(obj, p)
            p = XIRHXQ_Point(p);
            res = (obj.relation(p) == 0) && ((p - obj.s) * (p - obj.e) <= 0);
        end
        function res = parallel(obj1, obj2)
            mustBeA(obj1, 'XIRHXQ_Line');
            mustBeA(obj2, 'XIRHXQ_Line');
            res = ((obj1.e - obj1.s) ^ (obj2.e - obj2.s) == 0);
        end
        function res = is_the_same_line(obj1, obj2)
            mustBeA(obj1, 'XIRHXQ_Line');
            mustBeA(obj2, 'XIRHXQ_Line');
            res = (is_on_line(obj1, obj2.s) && is_on_line(obj1, obj2.e));
        end
        function res = line_cross_line(obj1, obj2)
            mustBeA(obj1, 'XIRHXQ_Line');
            mustBeA(obj2, 'XIRHXQ_Line');
            res = ~parallel(obj1, obj2);
        end
        function res = cross_point(obj1, obj2)
            mustBeA(obj1, 'XIRHXQ_Line');
            mustBeA(obj2, 'XIRHXQ_Line');
            a1 = (obj2.e - obj2.s) ^ (obj1.s - obj2.s);
            a2 = (obj2.e - obj2.s) ^ (obj1.e - obj2.s);
            if parallel(obj1, obj2)
                error('Not Crossing');
            end
            res = XIRHXQ_Point((obj1.s.x * a2 - obj1.e.x * a1) / (a2 - a1), (obj1.s.y * a2 - obj1.e.y * a1) / (a2 - a1));
        end
        function res = line_cross_seg(obj1, obj2)
            mustBeA(obj1, 'XIRHXQ_Line');
            mustBeA(obj2, 'XIRHXQ_Line');
            if is_the_same_line(obj1, obj2)
                error('Same Line');
            else
                res = (obj1.relation(obj2.s) * obj1.relation(obj2.e) < 0);
            end
        end
        function res = seg_cross_seg(obj1, obj2)
            mustBeA(obj1, 'XIRHXQ_Line');
            mustBeA(obj2, 'XIRHXQ_Line');
            if is_the_same_line(obj1, obj2)
                res = (obj1.relation_to_seg(obj2.s) * obj1.relation_to_seg(obj2.e) <= 0);
            elseif parallel(obj1, obj2)
                res = false;
            else
                cp = cross_point(obj1, obj2);
                res = obj1.is_on_seg(cp) && obj2.is_on_seg(cp);
            end
        end
        function res = dis_point_to_line(obj, p)
            p = XIRHXQ_Point(p);
            res = abs((p - obj.s) ^ (p - obj.e)) / obj.length;
        end
        function res = dis_point_to_seg(obj, p)
            p = XIRHXQ_Point(p);
            if (p - obj.s) * (obj.e - obj.s) < 0 || (p - obj.e) * (obj.s - obj.e) < 0
                res = min(p.distance(obj.s), p.distance(obj.e));
            else
                res = obj.dis_point_to_line(p);
            end
        end
        function res = dis_seg_to_seg(obj1, obj2)
            mustBeA(obj1, 'XIRHXQ_Line');
            mustBeA(obj2, 'XIRHXQ_Line');
            res = min(min(obj1.dis_point_to_seg(obj2.s), obj1.dis_point_to_seg(obj2.e)),...
                      min(obj2.dis_point_to_seg(obj1.s), obj2.dis_point_to_seg(obj1.e)));
        end
        function res = project_point(obj, p)
            p = XIRHXQ_Point(p);
            res = obj.s + (obj.e - obj.s) * ((obj.e - obj.s) * (p - obj.s)) / len2(obj.e - obj.s);
        end
        function res = symmetry_point(obj, p)
            mid = project_point(obj, p);
            res = mid * 2 - p;
        end
        function disp(obj)
            fprintf('A XIRHXQ_Line from (%.2f, %.2f) to (%.2f, %.2f)\n', obj.s.x, obj.s.y, obj.e.x, obj.e.y);
        end
    end
    methods(Static)
        function res = midean_line(p1, p2)
            p1 = XIRHXQ_Point(p1);
            p2 = XIRHXQ_Point(p2);
            res = XIRHXQ_Line((p1+p2)/2, angle(p2-p1) + pi/2);
        end
    end
end

File: cpp_file/XIRHXQ_Line.cpp
// Created by XIRHXQ on 12/29/21.

#include "utils.h"
#include "XIRHXQ_Line.h"

XIRHXQ_Line::XIRHXQ_Line() {
    s = XIRHXQ_Point(0, 0);
    e = XIRHXQ_Point(1, 0);
    adjust();
}

XIRHXQ_Line::XIRHXQ_Line(const XIRHXQ_Point &_s, const XIRHXQ_Point &_e) {
    s = _s;
    e = _e;
    adjust();
}

XIRHXQ_Line::XIRHXQ_Line(const XIRHXQ_Point &_s, double a) {
    s = _s;
    if (sgn(a - pi / 2) == 0){
        e = s + XIRHXQ_Point(0, 1);
    }
    else{
        e = s + XIRHXQ_Point(1, tan(a));
    }
}

XIRHXQ_Line::XIRHXQ_Line(double _a, double _b, double _c) {
    if (sgn(_a) == 0){
        s = XIRHXQ_Point(0, -_c / _b);
        e = XIRHXQ_Point(1, -_c / _b);
    }
    else if (sgn(_b) == 0){
        s = XIRHXQ_Point(-_c / _a, 0);
        e = XIRHXQ_Point(-_c / _a, 1);
    }
    else{
        s = XIRHXQ_Point(0, -_c / _b);
        e = XIRHXQ_Point(1, -(_c + _a) / _b);
    }
    adjust();
}

void XIRHXQ_Line::input(){
    s.input();
    e.input();
}

void XIRHXQ_Line::output() const {
    printf("A XIRHXQ_Line from (%.4lf, %.4lf) to (%.4lf, %.4lf)\n", s.x, s.y, e.x, e.y);
}

void XIRHXQ_Line::adjust() {
    if (s > e){
        XIRHXQ_Point tmp_point = s;
        s = e;
        e = tmp_point;
    }
}

double XIRHXQ_Line::len() {
    return s.distance_to(e);
}

double XIRHXQ_Line::len2() {
    return (s - e).len2();
}

double XIRHXQ_Line::angle() {
    return (e - s).angle();
}

int XIRHXQ_Line::relation(const XIRHXQ_Point &p) {
    int c = sgn((p - s) ^ (p - e));
    if (c < 0) return -1;
    else if (c > 0) return 1;
    else return 0;
}

int XIRHXQ_Line::relation_to_seg(const XIRHXQ_Point &p) {
    return sgn((p - s) ^ (e - s)) == 0 && sgn((p - s) * (p - e)) <= 0;
}

bool XIRHXQ_Line::is_on_line(const XIRHXQ_Point &p) {
    return relation(p) == 0;
}

bool XIRHXQ_Line::is_on_seg(const XIRHXQ_Point &p) {
    return relation_to_seg(p) == 0;
}

bool XIRHXQ_Line::is_parallel(const XIRHXQ_Line &l) {
    return sgn((e - s) ^ (l.e - l.s)) == 0;
}

bool XIRHXQ_Line::is_the_same_line(const XIRHXQ_Line &l) {
    return is_on_line(l.s) && is_on_line(l.e);
}

bool XIRHXQ_Line::line_cross_line(const XIRHXQ_Line &l) {
    return !is_parallel(l);
}

XIRHXQ_Point XIRHXQ_Line::cross_point(const XIRHXQ_Line &l) {
    double a1 = (l.e - l.s) ^ (s - l.s);
    double a2 = (l.e - l.s) ^ (e - l.s);
    return XIRHXQ_Point((s.x * a2 - e.x * a1) / (a2 - a1),
                        (s.y * a2 - e.y * a1) / (a2 - a1));
}

bool XIRHXQ_Line::line_cross_seg(const XIRHXQ_Line &l) {
    return (!is_parallel(l)) && (relation(l.s) * relation(l.e) < 0);
}

bool XIRHXQ_Line::seg_cross_seg(XIRHXQ_Line &l) {
    if (!is_parallel(l)) return false;
    XIRHXQ_Point cp = cross_point(l);
    return is_on_seg(cp) && l.is_on_seg(cp);
}

double XIRHXQ_Line::dis_line_to_point(const XIRHXQ_Point &p) {
    return fabs((p - s) ^ (e - s)) / len();
}

double XIRHXQ_Line::dis_seg_to_point(const XIRHXQ_Point &p) {
    if (sgn((p - s) * (e - s)) < 0 || sgn((p - e) * (s - e) < 0)){
        return fmin(s.distance_to(p), e.distance_to(p));
    }
    return dis_line_to_point(p);
}

double XIRHXQ_Line::dis_seg_to_seg(XIRHXQ_Line &l) {
    return fmin(fmin(dis_seg_to_point(l.s), dis_seg_to_point(l.e)),
               fmin(l.dis_seg_to_point(s), l.dis_seg_to_point(e)));
}

XIRHXQ_Point XIRHXQ_Line::project_point(const XIRHXQ_Point &p) {
    return s + (((e - s) * ((e - s) * (p - s))) / (e - s).len2() );
}

XIRHXQ_Point XIRHXQ_Line::symmetry_point(const XIRHXQ_Point &p) {
    XIRHXQ_Point q = project_point(p);
    return q * 2.0 - p;
}

XIRHXQ_Line median_line(XIRHXQ_Point &p, XIRHXQ_Point &q){
    return XIRHXQ_Line((p + q) / 2.0, p.angle_to(q) + pi / 2);
}